rust/macros/module.rs

```diff
@@ -270,23 +270,21 @@ pub(crate) fn module(ts: TokenStream) -> TokenStream {
             }}
 
             fn __init() -> core::ffi::c_int {{
-                match <{type_} as kernel::Module>::init(&THIS_MODULE) {{
-                    Ok(m) => {{
-                        unsafe {{
-                            __MOD = Some(m);
-                        }}
-                        return 0;
-                    }}
-                    Err(e) => {{
-                        return e.to_errno();
-                    }}
+                let initer = match <{type_} as kernel::InPlaceModule>::init(&THIS_MODULE) {{
+                    Ok(i) => i,
+                    Err(e) => return e.to_errno(),
+                }};
+
+                match unsafe {{ initer.__pinned_init(__MOD.as_mut_ptr()) }} {{
+                    Ok(m) => 0,
+                    Err(e) => e.to_errno(),
                 }}
             }}
 
             fn __exit() {{
                 unsafe {{
                     // Invokes `drop()` on `__MOD`, which should be used for cleanup.
-                    __MOD = None;
+                    __MOD.assume_init_drop();
```

Comment:
Does the `__exit()` function never get called, when `__init` returned non-zero?

Reply:
Correct. If `init` fails, `exit` is never called.